src/cargo/util/command_prelude.rs

```diff
@@ -1027,6 +1040,42 @@ pub fn lockfile_path(
     return Ok(Some(path));
 }
 
+fn get_bin_candidates() -> Vec<clap_complete::CompletionCandidate> {
+    get_bins_from_manifest()
+        .map(|bins| {
+            bins.into_iter()
+                .map(|bin| clap_complete::CompletionCandidate::new(bin))
+                .collect()
+        })
+        .unwrap_or(vec![])
+}
+
+fn get_bins_from_manifest() -> Option<Vec<String>> {
+    let manifest = get_manifest_with_table().ok()?;
+    let mut bins = vec![];
+
+    let bins_array = manifest.get("bin")?.as_array()?;
+
+    for bin in bins_array {
+        let name = bin.as_table()?.get("name")?.as_str()?;
+        bins.push(name.to_owned());
+    }
+
+    Some(bins)
+}
+
+fn get_manifest_with_table() -> CargoResult<toml::Table> {
+    let menifest = env!("CARGO_MANIFEST_DIR");
+    let menifest = Path::new(menifest).join("Cargo.toml");
+
+    let value = toml::from_str::<toml::Value>(&std::fs::read_to_string(menifest)?)?;
+
+    match value {
+        toml::Value::Table(table) => Ok(table),
+        _ => Err(anyhow!("Error parsing Cargo.toml")),
+    }
+}
```

Comment:
Most bins won't be found in `Cargo.toml`.

We should run logic like `cargo metadata` to load the workspace, enumerating targets and then filter the build targets for the ones that are bins.  This will then be easy to extend to examples, tests, benches, etc.